Add dockershim faq (#25360)
kubernetes/website: content/en/docs/setup/production-environment/dockershim.md

docker CLI commands in a way that isn’t repetitive] for system maintenance or
nested inside a container when building images will no longer work. For the
former, you can use the [`crictl`][cr] tool as a mostly drop-in replacement
and for the latter you can use newer container build options like [img], [buildah],
or [kaniko] that don’t require Docker.

[cr]: https://github.com/kubernetes-sigs/cri-tools
[img]: https://github.com/genuinetools/img
[buildah]: https://github.com/containers/buildah
[kaniko]: https://github.com/GoogleContainerTools/kaniko

For containerd in particular, you can start with their [documentation] to see what
configuration options are available as you migrate things over.

[documentaiton]: https://github.com/containerd/cri/blob/master/docs/registry.md

For instructions on how to use containerd and CRI-O with Kubernetes, see the
Kubernetes documentation on [Container Runtimes]

[Container Runtimes]: /docs/setup/production-environment/container-runtimes

### Why is dockershim being deprecated?
Maintaining dockershim has become a heavy burden on the Kubernetes maintainers.
The CRI standard was created to reduce this burden and allow smooth interoperability

> of different container runtimes, however Docker itself does not currently implement

Reviewer: runtimes. Docker itself doesn't currently implement CRI, thus the problem.

> CRI. Dockershim was always intended to be a temporary solution (hence the name).

Reviewer: "hence the name: shim"

Just being explicit

You can read more about the community discussion and planning in the 
[Dockershim Removal Kubernetes Enhancement Proposal][drkep]

[drkep]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1985-remove-dockershim

### What if I have more questions?

If you use a vendor-supported Kubernetes distribution, you can ask them about
upgrade plans for their products. For end-user questions, you can find our

> support volunteers on https://discuss.kubernetes.io/. 

Reviewer: For end-user questions, you can ask the community questions on https://discuss.kubernetes.io/.

---
title: Dockershim Deprecation FAQ
content_type: concept
aliases: [ '/dockershim']
weight: 40
---

This document goes over some frequently asked questions and their answers. For
more detail on the deprecation of Docker as a container runtime for Kubernetes
kubelets, and what that means, check out the blog post
[Don't Panic: Kubernetes and Docker](/blog/2020/12/02/dont-panic-kubernetes-and-docker/)

### Can I still use Docker in Kubernetes 1.20?

> Yes, the only thing changing in 1.20 is a single warning log printed at startup

Reviewer 2: should we mention that it's in kubelet startup, specifically? (and maybe link to the docs for kubelet?)

https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/

if using Docker as the runtime.

### When will dockershim be removed?

Given the impact of this change, we are using an extended deprecation timeline.
It will not be removed before Kubernetes 1.22, meaning the earliest release without
dockershim would be 1.23 in late 2021, approximately one year from now. We will
be working closely with vendors and other ecosystem groups to ensure a smooth
transition and will evaluate things as the situation evolves. 

### Are there examples of folks using other runtimes in production today that I can use to make the case to move forward with this change?

The [kind] project has been using containerd for some time and has seen an
improvement in stability for its use case. Kind and containerd are leveraged
multiple times every day to validate any changes to the Kubernetes codebase. Other
related projects follow a similar pattern as well, demonstrating the stability and
usability of other container runtimes.

> 

Reviewer 3: We can add a sentence around OpenShift 4.x using CRI-O in production since June 2019 as well here.

For other examples and references you can look at the adopters of containerd and
cri-o, two container runtimes under the Cloud Native Computing Foundation ([CNCF]).
- [containerd](https://github.com/containerd/containerd/blob/master/ADOPTERS.md)
- [CRI-O](https://github.com/cri-o/cri-o/blob/master/ADOPTERS.md)

[kind]: https://kind.sigs.k8s.io/
[CNCF]: https://cncf.io

### Will my existing Docker images still work?
Yes, the images produced from `docker build` will work with all CRI implementations.
All your existing images will still work exactly the same.

### What about private images?

Also yes. All CRI runtimes support the same pull secrets configuration used in
Kubernetes, either via the PodSpec or ServiceAccount.

### People keep referencing OCI, what is that?

OCI stands for the [Open Container Initiative], which standardized a lot of the
interfaces between container tools and technologies. They maintain a standard
specification for packaging container images (OCI image-spec) and running containers

> (OCI runtime-spec). The CRI builds on these low-level specifications to provide

Reviewer 3: We should also mention runc which is being maintained as an implementation of the runtime spec within OCI and is typically the low level runtimes that is the default in containerd and CRI-O as it was in docker as well.

an end-to-end standard for managing containers.

[Open Container Initative]: https://opencontainers.org/about/overview/

### Are Docker and containers the same thing?

Docker popularized the Linux containers pattern and has been instrumental in
developing the underlying technology, however containers in linux have existed
for a long time. The container ecosystem has grown to be much broader than just
Docker. Standards like OCI and CRI have helped many tools grow and thrive in our
ecosystem, some replacing aspects of Docker while others enhance existing
functionality.

### Which CRI implementation should I use?

That’s a complex question and it depends on a lot of factors. If Docker is
working for you, moving to containerd should be a relatively easy swap and

> has have strictly better performance and le overhead. However we encourage you

Reviewer 3: le -> less

to explore all the options from the [CNCF landscape] in case another would be an
even better fit for your environment.

[CNCF landscape]: https://landscape.cncf.io/category=container-runtime&format=card-mode&grouping=category

### What should I look out for when changing CRI implementations?

While the underlying containerization code is the same between Docker and most
CRIs (including containerd), there are a few differences around the edges. Some
common things to consider when migrating are:
- Logging configuration
- Runtime resource limitations

> - Node provisioning scripts that require Docker or the controlsocket

Reviewer 3: .. call docker or use docker via it's control socket

- Kubectl plugins that require docker CLI or the control socket
- Kubernetes tools that require direct access to Docker (e.g. kube-imagepuller)
- Configuration of functionality like `registry-mirrors` and insecure registries 
- Other support scripts or daemons that expect docker to be available and are run

>   outside of Kubernetes (e.g. monitoring or security agents)

Reviewer 3: We can another bullet for how devices are enabled in runtimes so users that want GPUs or other special hardware should look up the documentation on how to enable those when switching runtimes.

If you use Kubernetes resource requests/limits or file-based log collection
DaemonSets then they will continue to work the same, but if you’ve customized
your dockerd configuration, you’ll need to adapt that for your new container
runtime where possible.

Another thing to look out for is anything expecting to run [ed: how to say
docker CLI commands in a way that isn’t repetitive] for system maintenance or
nested inside a container when building images will no longer work. For the
former, you can use the [`crictl`][cr] tool as a mostly drop-in replacement
and for the latter you can use newer container build options like [img], [buildah],
or [kaniko] that don’t require Docker.

[cr]: https://github.com/kubernetes-sigs/cri-tools
[img]: https://github.com/genuinetools/img
[buildah]: https://github.com/containers/buildah
[kaniko]: https://github.com/GoogleContainerTools/kaniko

For containerd in particular, you can start with their [documentation] to see what
configuration options are available as you migrate things over.

> [documentaiton]: https://github.com/containerd/cri/blob/master/docs/registry.md

Reviewer 3: typo: documentation

For instructions on how to use containerd and CRI-O with Kubernetes, see the
Kubernetes documentation on [Container Runtimes]

[Container Runtimes]: /docs/setup/production-environment/container-runtimes

### Why is dockershim being deprecated?

Maintaining dockershim has become a heavy burden on the Kubernetes maintainers.
The CRI standard was created to reduce this burden and allow smooth interoperability
of different container runtimes, however runtimes. Docker itself doesn't currently
implement CRI, thus the problem. Dockershim was always intended to be a temporary
solution (hence the name: shim). You can read more about the community discussion
and planning in the [Dockershim Removal Kubernetes Enhancement Proposal][drkep]

[drkep]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1985-remove-dockershim

> 

Reviewer 3: Newer features such as cgroups v2 and user namespaces are being implemented in the new CRI runtimes and cutting away from dockershim allows us to move faster in supporting new kernel features for containerization.

---
title: Dockershim Deprecation FAQ
content_type: concept
aliases: [ '/dockershim']
weight: 40
---

> This document goes over some frequently asked questions and their answers. For

Reviewer 4: This document reviews some frequently asked questions about [what]?

> more detail on the deprecation of Docker as a container runtime for Kubernetes

Reviewer 2: there's already a blog post verison https://kubernetes.io/blog/2020/12/02/dont-panic-kubernetes-and-docker/

> more detail on the deprecation of Docker as a container runtime for Kubernetes

Author: its being migrated to a blog post now. 👍

> more detail on the deprecation of Docker as a container runtime for Kubernetes

Reviewer 4: Yes, I understand. How will readers know to look in this section for important information?
It would be good to broadcast widely and often (release notes?).

kubelets, and what that means, check out the blog post

> [Don't Panic: Kubernetes and Docker](/blog/2020/12/02/dont-panic-kubernetes-and-docker/)

Reviewer 5: ```suggestion
[Don't Panic: Kubernetes and Docker](/blog/2020/12/02/dont-panic-kubernetes-and-docker/).
```

### Can I still use Docker in Kubernetes 1.20?

Yes, the only thing changing in 1.20 is a single warning log printed at [kubelet]

> startup if using Docker as the runtime.

Reviewer 4: nit: Could you write the links as `[kubelet](/docs/reference/command-line-tools-reference/kubelet/)`

[kubelet]: /docs/reference/command-line-tools-reference/kubelet/

### When will dockershim be removed?

Given the impact of this change, we are using an extended deprecation timeline.
It will not be removed before Kubernetes 1.22, meaning the earliest release without

> dockershim would be 1.23 in late 2021, approximately one year from now. We will

Reviewer 6: I would remove the "approximately one year from now" because this won't have a date stamp on it.

Reviewer 7: Agreed, especially as we're discussing changes to release cadence.

be working closely with vendors and other ecosystem groups to ensure a smooth
transition and will evaluate things as the situation evolves. 

> ### Are there examples of folks using other runtimes in production today that I can use to make the case to move forward with this change?

Reviewer 4: Can you reduce the text in this heading? I have not viewed the rendered page, but this heading will display in the right hand TOC and the text will wrap over multiple lines?

The [kind] project has been using containerd for some time and has seen an
improvement in stability for its use case. Kind and containerd are leveraged

> multiple times every day to validate any changes to the Kubernetes codebase. Other

Reviewer 7: Maybe we want to explicitly say that it's part of release signal?

Author: I don't think most end users will really get "release signal"

Reviewer 7: Something along the lines of "we check this before we released Kubernetes".

i.e., it's Community-validated for the artifacts we produce is a stronger message over it only being "leveraged".

related projects follow a similar pattern as well, demonstrating the stability and
usability of other container runtimes. As an example, OpenShift 4.x has been
using the [CRI-O] runtime in production since June 2019.

For other examples and references you can look at the adopters of containerd and
cri-o, two container runtimes under the Cloud Native Computing Foundation ([CNCF]).
- [containerd](https://github.com/containerd/containerd/blob/master/ADOPTERS.md)
- [CRI-O](https://github.com/cri-o/cri-o/blob/master/ADOPTERS.md)

[CRI-O]: https://cri-o.io/
[kind]: https://kind.sigs.k8s.io/
[CNCF]: https://cncf.io

> ### Will my existing Docker images still work?

Reviewer 5: ```suggestion
### Will my existing Docker images still work?

```
header formatting breaks otherwise

Yes, the images produced from `docker build` will work with all CRI implementations.
All your existing images will still work exactly the same.

### What about private images?

Also yes. All CRI runtimes support the same pull secrets configuration used in
Kubernetes, either via the PodSpec or ServiceAccount.

### People keep referencing OCI, what is that?

> OCI stands for the [Open Container Initiative], which standardized a lot of the

Reviewer 5: ```suggestion
OCI stands for the [Open Container Initiative], which standardized many of the
```

interfaces between container tools and technologies. They maintain a standard
specification for packaging container images (OCI image-spec) and running containers
(OCI runtime-spec). They also maintain an actual implementation of the runtime-spec
in the form of [runc], which is the underlying default runtime for both
[containerd] and [CRI-O]. The CRI builds on these low-level specifications to
provide an end-to-end standard for managing containers.

[Open Container Initative]: https://opencontainers.org/about/overview/
[runc]: https://github.com/opencontainers/runc
[containerd]: https://containerd.io/

### Are Docker and containers the same thing?

Docker popularized the Linux containers pattern and has been instrumental in

> developing the underlying technology, however containers in linux have existed

Reviewer 5: ```suggestion
developing the underlying technology, however containers in Linux have existed
```

for a long time. The container ecosystem has grown to be much broader than just
Docker. Standards like OCI and CRI have helped many tools grow and thrive in our
ecosystem, some replacing aspects of Docker while others enhance existing
functionality.

### Which CRI implementation should I use?

That’s a complex question and it depends on a lot of factors. If Docker is
working for you, moving to containerd should be a relatively easy swap and
has have strictly better performance and less overhead. However we encourage you
to explore all the options from the [CNCF landscape] in case another would be an
even better fit for your environment.

[CNCF landscape]: https://landscape.cncf.io/category=container-runtime&format=card-mode&grouping=category

### What should I look out for when changing CRI implementations?

While the underlying containerization code is the same between Docker and most
CRIs (including containerd), there are a few differences around the edges. Some

> common things to consider when migrating are:

Reviewer 5: ```suggestion
common things to consider when migrating are:

```

(again, minor md formatting issue)

- Logging configuration
- Runtime resource limitations
- Node provisioning scripts that call docker or use docker via it's control socket
- Kubectl plugins that require docker CLI or the control socket
- Kubernetes tools that require direct access to Docker (e.g. kube-imagepuller)
- Configuration of functionality like `registry-mirrors` and insecure registries 
- Other support scripts or daemons that expect docker to be available and are run
  outside of Kubernetes (e.g. monitoring or security agents)
- GPUs or special hardware and how they integrate with your runtime and Kubernetes

If you use Kubernetes resource requests/limits or file-based log collection
DaemonSets then they will continue to work the same, but if you’ve customized
your dockerd configuration, you’ll need to adapt that for your new container
runtime where possible.

> Another thing to look out for is anything expecting to run [ed: how to say

Reviewer 4: This is hard to parse:
Another thing to look out for is anything expecting to run [ed: how to say

docker CLI commands in a way that isn’t repetitive] for system maintenance or
nested inside a container when building images will no longer work. For the

> former, you can use the [`crictl`][cr] tool as a mostly drop-in replacement

Reviewer 4: nit: remove or reword "as a mostly drop-in"

Reviewer 4: nit: The links.

and for the latter you can use newer container build options like [img], [buildah],
or [kaniko] that don’t require Docker.

[cr]: https://github.com/kubernetes-sigs/cri-tools
[img]: https://github.com/genuinetools/img
[buildah]: https://github.com/containers/buildah
[kaniko]: https://github.com/GoogleContainerTools/kaniko

For containerd in particular, you can start with their [documentation] to see what

> configuration options are available as you migrate things over.

Reviewer 4: nit: could remove "in particular" and make the link text more descriptive

[documentation]: https://github.com/containerd/cri/blob/master/docs/registry.md

For instructions on how to use containerd and CRI-O with Kubernetes, see the
Kubernetes documentation on [Container Runtimes]

[Container Runtimes]: /docs/setup/production-environment/container-runtimes

> ### Why is dockershim being deprecated?

Reviewer 5: Move this question to the very top of the list.

In short: order this based on **relevance to the casual passer-by**.  

If I know nothing about this change other than some scary twitter thread by a Senior Clout Engineer, I'd want the 5 W's (Who/What/When/Where/**Why**) answered before I dove into implementation details, picking a replacement, etc.

Maintaining dockershim has become a heavy burden on the Kubernetes maintainers.
The CRI standard was created to reduce this burden and allow smooth interoperability

> of different container runtimes, however runtimes. Docker itself doesn't currently

Reviewer 5: ```suggestion
of different container runtimes. Docker itself doesn't currently
```

side note: "However, runtimes" is my new objection to any Kubernetes enhancement. I don't comment on many keps tho so maybe @justaugustus can do the honors 😂

> implement CRI, thus the problem. Dockershim was always intended to be a temporary

Reviewer 5: ```suggestion
implement CRI, thus the problem. 

Dockershim was always intended to be a temporary
```

New paragraph here.
